### ISSUE TITLE
'query' keyword automtically set at the beginning of the query string

### ISSUE THREAD

**Issue by AndoniLarz** [mghoneimy/php-graphql-client#8]
Hi, I'm working on calling a GraphQL API and I found this package was very useful, thanks for making it !

However, I encountered a little problem :
It seems like the query is automatically completed with a `query` keyword at the beginning of my query string as in this snippet : 
`query {
content {
product(id: xx) {
id
}
}
}
`
Is it because of some GraphQL rule ?
In fact it blocks me from calling the API since my query should begin with the `content` keyword.

Would you be able to give choice about that automatic keyword insertion ?

**Comment by mghoneimy** [comment]
Hello @AndoniLarz , first of all, glad that my package was of help to you :)

As for you question, let me answer it in 2 parts.
First of all, GraphQL supports two main types of operations that serve as an "entry point" to the graph. Each request to a GraphQL server represents an "operation", and this operation has a type, which in the case of a query would be "query" and in the case of mutation would be "mutation". Mutations must have an operation type, but if not operation type is provided, then the operation is assumed to be a query.
For being explicit, I append the "query" keyword to all query operations created by this package, which should work on all GraphQL servers without a problem.

I highly doubt that the GraphQL server is not working because the query needs to start with the keyword `content`. This requires that content becomes a root operation type, which is impossible by GraphQL standards https://graphql.github.io/graphql-spec/June2018/#sec-Root-Operation-Types.

The query either starts with the `query` keyword followed by braces containing the query body, or the braces and the query body directly like this:
```
{
    content {
        product(id: xx) {
            id
        }
    }
}
```
or with the query word like this:
```
query {
    content {
        product(id: xx) {
            id
        }
    }
}
```

As for the second part of the question of I would give choice to the caller. I can, but I wouldn't. GraphQL supports only three root operation type: query, mutation, and subscription. I've already covered two of them and intend to cover the third soon. But there shouldn't be any cases where server declare their own root operation types, if they abide to Facebook's GraphQL standards.

Maybe if you show me the error you get from the server I can help you with it, but I'm pretty sure that it can't be the query keyword, or else, the server would be violating the GraphQL standards.

**Comment by AndoniLarz** [comment]
Thank you for this fast answer and all the explanations ! As you may have noticed I'm not very familiar with GraphQL, so I didn't know where to look at to find an RFC-like document as you provided.
I'll look into it tomorrow at work, but I guess the API should be implemented following those standards.

I'll keep you updated as soon as I have more info !

**Comment by mghoneimy** [comment]
Sounds great!

**Comment by AndoniLarz** [comment]
Okay so I figured out what the problem was :
I am calling the GraphQL API in order to check if an entity already exists in the DB. Logically, when it doesn't exist, the API returns an `errors` block that explains the error and a `data` block that contains an array with only one `null` value.

I needed to get the `data` block, so I added a $data member and a getData method to be able to get this block.
I'm going to make a PR, so that we can discuss on the PR about this decision.

**Comment by AndoniLarz** [comment]
Here it is : https://github.com/mghoneimy/php-graphql-client/pull/9

**Comment by mghoneimy** [comment]
@AndoniLarz 
Hmmm I understand what you want to do now, but I thought no one would want to retrieve the `data` when `errors` are returned. I supported this implementation assumption by the fact that `data` key _should_ be `null` when `errors` are set and nobody would be interested in retrieving this `null` value.

This is how it is stated by the GraphQL spec at https://graphql.github.io/graphql-spec/June2018/#sec-Data :
```"If an error was encountered during the execution that prevented a valid response, the data entry in the response should be null."```
This is also aligned with the behavior you see from the GraphQL server you're working with.

That's why when `errors` key is encountered during parsing the response a `QueryException` is thrown which prevents you from accessing the `data` key in the response, which should always be null anyways.

Can you explain to me why you want to retrieve this value since it's always going to be `null` if set?

Meanwhile, I'll review your pull request.

**Comment by AndoniLarz** [comment]
@mghoneimy Ok I understand your point, which is totally valid, but apart from my own interest I found that the `data` key may not be `null` in some cases, as shown here : https://graphql.github.io/graphql-spec/June2018/#example-90475
So it can be useful to retrieve this part of the response anyways.

As of my case of getting `data` even if it's `null` : it was a total misuse since I should parse the `message` key of the `errors` part, instead of relying on a `null` value for the `product` key.
I was using it to make sure the `product` I'm looking for is not present in the DB.

TL;DR : I think the PR can still be useful to some, considering the case of the documentation I linked, but I need to refactor how I check the existence of the product in the DB by parsing the `message` key.

Do you agree ?

**Comment by mghoneimy** [comment]
@AndoniLarz I think that you have a valid caze with that example, although I haven't actually come across an API that exhibits this behavior yet.

Let's work on finalizing this pull request to merge and release it.
